moveit_planners/ompl/ompl_interface/include/moveit/ompl_interface/parameterization/model_based_state_space.h

```diff
@@ -86,84 +87,148 @@ class ModelBasedStateSpace : public ompl::base::StateSpace
+      });
     }
 
-    void markInvalid(double d)
+    void markValid() const
     {
-      distance = d;
-      flags |= GOAL_DISTANCE_KNOWN;
-      markInvalid();
+      setFlag(VALIDITY_KNOWN | VALIDITY_TRUE);
     }
 
-    void markInvalid()
+    void markInvalid(double d) const
     {
-      flags &= ~VALIDITY_TRUE;
-      flags |= VALIDITY_KNOWN;
+      modifyCache([d](AtomicCache& desired) {
+        desired.distance = d;
+        desired.flags &= ~VALIDITY_TRUE;
+        desired.flags |= (VALIDITY_KNOWN | GOAL_DISTANCE_KNOWN);
+      });
     }
 
-    bool isValidityKnown() const
+    void markInvalid() const
     {
-      return flags & VALIDITY_KNOWN;
+      modifyCache([](AtomicCache& desired) {
+        desired.flags &= ~VALIDITY_TRUE;
+        desired.flags |= VALIDITY_KNOWN;
+      });
     }
 
     void clearKnownInformation()
     {
-      flags = 0;
+      modifyCache([](AtomicCache& desired) { desired.flags = 0; });
     }
 
-    bool isMarkedValid() const
+    void markStartState()
     {
-      return flags & VALIDITY_TRUE;
+      setFlag(IS_START_STATE);
     }
 
-    bool isGoalDistanceKnown() const
+    void markGoalState()
     {
-      return flags & GOAL_DISTANCE_KNOWN;
+      setFlag(IS_GOAL_STATE);
     }
 
-    bool isStartState() const
+    int flags() const
     {
-      return flags & IS_START_STATE;
+      return atomic_cache.load().flags;
     }
-
-    bool isGoalState() const
+    void setFlag(int flag) const
     {
-      return flags & IS_GOAL_STATE;
+      modifyCache([flag](AtomicCache& desired) { desired.flags |= flag; });
     }
-
-    bool isInputState() const
+    void clearflag(int flag) const
     {
-      return flags & (IS_START_STATE | IS_GOAL_STATE);
+      modifyCache([flag](AtomicCache& desired) { desired.flags &= ~flag; });
     }
-
-    void markStartState()
+    int tag() const
     {
-      flags |= IS_START_STATE;
+      return atomic_cache.load().tag;
+    }
+    void setTag(int tag)
+    {
+      modifyCache([tag](AtomicCache& desired) { desired.tag = tag; });
+    }
+    AtomicCache getCache() const
+    {
+      return atomic_cache.load();
+    }
+    void setCache(const AtomicCache& cache)
+    {
+      atomic_cache.store(cache);
     }
 
-    void markGoalState()
+  protected:
+    mutable std::atomic<AtomicCache> atomic_cache;
+    /**
+     * \brief Helper function to modify the cache in a compare-and-swap loop.
+     *
+     * The Lambda will be called at least one time.
+     * It takes a reference to a copy of the cache and should update it
+     * (e.g. set a bit in the flags). The copy is then written back,
+     * but only if the original cache was left untouched.
+     *
+     * \param func Lambda which takes exactly one parameter, an AtomicCache Reference.
+     *
+     */
+    template <class Func>
+    void modifyCache(Func&& func) const
     {
-      flags |= IS_GOAL_STATE;
+      AtomicCache desired, expected = atomic_cache.load();
+      do
```

Comment:
I do not understand why the loop is necessary. Could you maybe point to an article or doc page I can read to understand this? :)

Reply:
This is a compare-and-swap loop, for atomic Read-Modify-Write. Basically you make two copies of the atomic variable and modify one of them. `<atomic>` Then, you compare the unaltered copy with the atomic variable and if they match, you swap the atomic it with the modified variable.`</atomic>`. If the atomic variable has been altered, you have to start over. The CAS can also fail spuriously, so a loop is always needed.
In our case, we use it to set the flag and the distance at the same time. If only `flags` were atomic, thread B could change the `distance` between reading `flags` and `distance` in thread A.